pennylane/gradients/gradient_transform.py

```diff
@@ -0,0 +1,178 @@
+              processing will be computed and included. When evaluated, the
+              returned Jacobian will be with respect to the QNode arguments.
+
+            - If ``False``, any internal QNode classical processing will be
+              **ignored**. When evaluated, the returned Jacobian will be with
+              respect to the **gate** arguments, and not the QNode arguments.
+
+    Supported gradient transforms must be of the following form:
+
+    .. code-block:: python
+
+        @quantum_gradient
+        def my_custom_gradient(tape, argnum=None, **kwargs):
+            ...
+            return gradient_tapes, processing_fn
+
+    where:
+
+    - ``tape`` (*QuantumTape*): the input quantum tape to compute the gradient of
+
+    - ``argnum`` (*int* or *list[int]* or *None*): Which trainable parameters of the tape
+      to differentiate with respect to. If not provided, the derivatives with respect to all
+      trainable indices of the tape should be returned (``tape.trainable_params``).
+
+    - ``gradient_tapes`` (*List[QuantumTape]*): is a list of output tapes to be evaluated.
+      If this list is empty, no quantum evaluations will be made.
+
+    - ``processing_fn`` is a processing function to be applied to the output of the evaluated
+      ``gradient_tapes``. It should accept a list of numeric results with length ``len(gradient_tapes)``,
+      and return the Jacobian matrix.
+
+    Once defined, the quantum gradient transform can be used as follows:
+
+    >>> gradient_tapes, processing_fn = my_custom_gradient(tape, *gradient_kwargs)
+    >>> res = execute(tapes, dev, interface="autograd", gradient_fn=qml.gradients.param_shift)
+    >>> jacobian = processing_fn(res)
+
+    Alternatively, gradient transforms can be applied directly to QNodes,
+    in which case the execution is implicit:
+
+    >>> fn = my_custom_gradient(qnode, *gradient_kwargs)
+    >>> fn(weights) # transformed function takes the same arguments as the QNode
+    1.2629730888100839
+
+    .. note::
+
+        The input tape might have parameters of various types, including
+        NumPy arrays, JAX DeviceArrays, and TensorFlow and PyTorch tensors.
+
+        If the gradient transform is written in a autodiff-compatible manner, either by
+        using a framework such as Autograd or TensorFlow, or by using ``qml.math`` for
+        tensor manipulation, than higher-order derivatives will also be supported.
+
+        Alternatively, you may use the ``tape.unwrap()`` context manager to temporarily
+        convert all tape parameters to NumPy arrays and floats:
+
+        >>> with tape.unwrap():
+        ...     params = tape.get_parameters()  # list of floats
+    """
+
+    def __init__(self, transform_fn, expand_fn=gradient_expand, differentiable=True, hybrid=True):
+        self.hybrid = hybrid
+        super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
+
+    def qnode_execution_wrapper(self, qnode, targs, tkwargs):

```

Comment:
Have just skimmed this section in my first pass, will take a closer look in a subsequent one.